seastar/src/submit_to.rs

```diff
@@ -0,0 +1,122 @@
+use crate as seastar;
+use ffi::*;
+use std::future::Future;
+
+use crate::cxx_async_local_future::IntoCxxAsyncLocalFuture;
+
+#[cxx::bridge]
+mod ffi {
+    #[namespace = "seastar_ffi"]
+    unsafe extern "C++" {
+        type VoidFuture = crate::cxx_async_futures::VoidFuture;
+    }
+
+    #[namespace = "seastar_ffi::submit_to"]
+    unsafe extern "C++" {
+        include!("seastar/src/submit_to.hh");
+
+        unsafe fn submit_to(
+            shard_id: u32,
+            closure: *mut u8,
+            caller: unsafe fn(*mut u8) -> VoidFuture,
+        ) -> VoidFuture;
+    }
+}
+
+/// Convert the pointer to Box<Func> and then call it, consuming it in the process.
+fn fn_once_caller<Func>(raw_func: *mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    unsafe { Box::from_raw(raw_func as *mut Func)() }
+}
+
+/// A helper function. We need to be able to name the type of a named closure.
+/// There is no `typeof` in Rust, so this is achieved by using a dummy paramter.
+const fn get_fn_once_caller<Func>(_: &Func) -> fn(*mut u8) -> VoidFuture
+where
+    Func: FnOnce() -> VoidFuture,
+{
+    fn_once_caller::<Func>
+}
+
+/// Free a pointer. We need to call this function in the case of an error.
+fn fn_once_dropper<Func>(raw_func: *mut u8) {
+    unsafe {
+        let _ = Box::from_raw(raw_func as *mut Func);
+    }
+}
+
+const fn get_fn_once_dropper<Func>(_: &Func) -> fn(*mut u8) {
+    fn_once_dropper::<Func>
+}
```

Comment:
I anticipate that those functions might be useful in other contexts, e.g. there is some overlap with what the timers are doing. Please move them to a new (internal) module `ffi_utils`.